File: mobile/src/main/java/xyz/mechenbier/circuittester/MainActivity.java
package xyz.mechenbier.circuittester;

import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.content.SharedPreferences;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.os.IBinder;
import android.support.v4.app.NotificationCompat;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.View;
import android.widget.ImageView;
import android.widget.RadioButton;
import android.widget.ToggleButton;
import com.google.android.gms.ads.AdView;
import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.MobileAds;
import com.google.firebase.analytics.FirebaseAnalytics;

import xyz.mechenbier.circuittester.PowerStateService.LocalBinder;

public class MainActivity extends AppCompatActivity {

    private PowerStateService mService;
    private boolean mBound = false;
    private UiUpdateReceiver mUiUpdateReceiver;
    private static MainActivity mActivityInstance;
    private FirebaseAnalytics mFirebaseAnalytics;
    private static final int mNotificationID = 5000;
    private static final int RATING_DAYS_UNTIL_PROMPT = 3;
    private static final int RATING_LAUNCHES_UNTIL_PROMPT = 3;

    //Defines callbacks for service binding, passed to bindService()
    private ServiceConnection mConnection = new ServiceConnection() {

        @Override
        public void onServiceConnected(ComponentName className, IBinder service) {
            // We've bound to LocalService, cast the IBinder and get LocalService instance
            LocalBinder binder = (LocalBinder)service;
            mService = binder.getService();
            mBound = true;
        }

        @Override
        public void onServiceDisconnected(ComponentName arg0) {
            mBound = false;
        }
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        mActivityInstance = this;

        mFirebaseAnalytics = FirebaseAnalytics.getInstance(this);

        MobileAds.initialize(this, BuildConfig.AdMobAppApiKey);
        AdView adView = (AdView) findViewById(R.id.adView);
        AdRequest adRequest = new AdRequest.Builder()
                .addTestDevice(AdRequest.DEVICE_ID_EMULATOR)
                .build();
        adView.loadAd(adRequest);

        initializeRatingPrompt();

        startService();
    }

    @Override
    protected void onStart() {
        super.onStart();
        bindService();
    }

    @Override
    protected void onResume() {
        if (mUiUpdateReceiver == null) {
            mUiUpdateReceiver = new UiUpdateReceiver();
        }
        android.content.IntentFilter intentFilter = new android.content.IntentFilter(getString(R.string.intent_charging_name));
        registerReceiver(mUiUpdateReceiver, intentFilter);

        super.onResume();
    }

    @Override
    protected void onPause() {
        if (mUiUpdateReceiver != null) {
            unregisterReceiver(mUiUpdateReceiver);
        }
        super.onPause();
    }

    @Override
    protected void onStop() {
        super.onStop();
        unbindService();
    }

    @Override
    protected void onDestroy() {
        stopService();
        super.onDestroy();
    }


    public static MainActivity getInstance() {
        return mActivityInstance;
    }


    public void createNotification() {
        createNotificationChannel(this);

        Intent intent = new Intent(this, MainActivity.class);
        PendingIntent contentIntent = PendingIntent
                .getActivity(this, 0, intent, PendingIntent.FLAG_UPDATE_CURRENT);

        NotificationCompat.Builder notificationBuilder = new NotificationCompat.Builder(this, getString(R.string.notification_channel_id))
                .setDefaults(NotificationCompat.DEFAULT_ALL)
                .setSmallIcon(R.drawable.bolt)
                .setContentTitle(getString(R.string.app_name))
                .setContentText(getString(R.string.notification_details))
                .setWhen(System.currentTimeMillis())
                .setOngoing(true)
                .setContentIntent(contentIntent);

        NotificationManager notificationManager = (NotificationManager)this
                .getSystemService(Context.NOTIFICATION_SERVICE);

        notificationManager.notify(mNotificationID, notificationBuilder.build());
    }

    private void createNotificationChannel(Context context) {
        // we don't need to do this for devices running Android O or lower but NotificationChannel is not in appcompat so we have to check the version
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            NotificationManager notificationManager = (NotificationManager)context.getSystemService(Context.NOTIFICATION_SERVICE);
            NotificationChannel notificationChannel = new NotificationChannel(getString(R.string.notification_channel_id), getString(R.string.notification_channel_name), NotificationManager.IMPORTANCE_DEFAULT);
            notificationChannel.setDescription(getString(R.string.notification_channel_description));
            notificationManager.createNotificationChannel((notificationChannel));
        }
    }

    public void dismissNotification() {
        NotificationManager notificationManager = (NotificationManager)this
                .getSystemService(Context.NOTIFICATION_SERVICE);
        notificationManager.cancel(mNotificationID);
    }


    public void bindService() {
        // Bind to LocalService
        Intent intent = new Intent(this, PowerStateService.class);
        bindService(intent, mConnection, Context.BIND_AUTO_CREATE);
    }

    private void unbindService() {
        try {
            unbindService(mConnection);
        } catch (Exception e) {
            Log.e("SERVICE", "Issue unbinding the service");
        }
        mBound = false;
    }

    public void startService(View view) {
        startService();
        bindService();
    }

    public void startService() {
        Intent intent = new Intent(this, PowerStateService.class);
        startService(intent);
        createNotification();
    }

    public void stopService(View view) {
        stopService();
    }

    private void stopService() {
        unbindService();
        dismissNotification();
        Intent intent = new Intent(this, PowerStateService.class);
        stopService(intent);
    }


    public void onRadioButtonClicked(View view) {
        boolean isChecked = ((RadioButton)view).isChecked();

        // Check which radio button was clicked
        switch (view.getId()) {
            case R.id.rb_sound_when_powered:
                setSoundOnPowered(!isChecked);
                break;
            case R.id.rb_sound_when_not_powered:
                setSoundOnPowered(isChecked);
                break;
        }
    }

    private void setSoundOnPowered(boolean checked) {
        if (mBound) {
            mService.setSoundOnPowered(checked);
        }
    }


    public void onMuteToggleButtonClicked(View view) {
        boolean isChecked = ((ToggleButton)view).isChecked();
        setMute(isChecked);
    }

    private void setMute(boolean muted) {
        if (mBound) {
            mService.setMute(muted);
        }
    }


    public void setChargingIconColor(boolean isCharging) {
        ImageView image = (ImageView) findViewById(R.id.image_powerstate);
        if (isCharging) {
            image.setColorFilter(getResources().getColor(R.color.image_charging));
        } else {
            image.setColorFilter(getResources().getColor(R.color.image_not_charging));
        }
    }


    private void initializeRatingPrompt() {
        SharedPreferences preferences = this.getSharedPreferences(getString(R.string.preference_shared_key_name_rating), 0);

        if (preferences.getBoolean(getString(R.string.preference_name_rating_dont_show_rating_prompt), false)) {
            return;
        }

        SharedPreferences.Editor preferencesEditor = preferences.edit();

        // increment launch counter
        long launch_count = preferences.getLong(getString(R.string.preference_name_rating_launch_count), 0) + 1;
        preferencesEditor.putLong(getString(R.string.preference_name_rating_launch_count), launch_count);

        // get date of first launch
        long date_first_launch = preferences.getLong(getString(R.string.preference_name_rating_launch_date), 0);
        if (date_first_launch == 0) {
            date_first_launch = System.currentTimeMillis();
            preferencesEditor.putLong(getString(R.string.preference_name_rating_launch_date), date_first_launch);
        }

        preferencesEditor.apply();

        // wait at least n days before opening
        if (launch_count >= RATING_LAUNCHES_UNTIL_PROMPT) {
            if (System.currentTimeMillis() >= date_first_launch + (RATING_DAYS_UNTIL_PROMPT * 24 * 60 * 60 * 1000)) {
                setVisibility(R.id.rating_frame_layout, View.VISIBLE);
                setVisibility(R.id.text_warning, View.INVISIBLE);
            }
        }
    }

    private void setVisibility(int id, int visibility) {
        View view = findViewById(id);
        if (view != null) {
            view.setVisibility(visibility);
        }
    }

    // when a user clicks the "not enjoying" rating button, hide the enjoyment prompts and show the feedback prompts
    public void onRatingNotEnjoyingButtonClicked(View view){
        setVisibility(R.id.enjoying_circuit_tester, View.INVISIBLE);
        setVisibility(R.id.button_enjoying_no, View.INVISIBLE);
        setVisibility(R.id.button_enjoying_yes, View.INVISIBLE);
        setVisibility(R.id.feedback_circuit_tester, View.VISIBLE);
        setVisibility(R.id.button_feedback_no, View.VISIBLE);
        setVisibility(R.id.button_feedback_yes, View.VISIBLE);
    }

    // when a user clicks the "yes enjoying" rating button, hide the enjoyment prompts and show the rating prompts
    public void onRatingYesEnjoyingButtonClicked(View view){
        setVisibility(R.id.enjoying_circuit_tester, View.INVISIBLE);
        setVisibility(R.id.button_enjoying_no, View.INVISIBLE);
        setVisibility(R.id.button_enjoying_yes, View.INVISIBLE);
        setVisibility(R.id.rate_circuit_tester, View.VISIBLE);
        setVisibility(R.id.button_rating_no, View.VISIBLE);
        setVisibility(R.id.button_rating_yes, View.VISIBLE);
    }

    // when a user clicks the "no feedback" rating button, hide the rating prompt
    public void onRatingNoFeedbackButtonClicked(View view){
        SharedPreferences preferences = this.getSharedPreferences(getString(R.string.preference_shared_key_name_rating), 0);
        SharedPreferences.Editor preferencesEditor = preferences.edit();
        preferencesEditor.putBoolean(getString(R.string.preference_name_rating_dont_show_rating_prompt), true);
        preferencesEditor.apply();
        setVisibility(R.id.rating_frame_layout, View.INVISIBLE);
        setVisibility(R.id.text_warning, View.VISIBLE);
    }

    // when a user clicks the "yes feedback" rating button, hide the prompt and create an email intent
    public void onRatingYesFeedbackButtonClicked(View view){
        SharedPreferences preferences = this.getSharedPreferences(getString(R.string.preference_shared_key_name_rating), 0);
        SharedPreferences.Editor preferencesEditor = preferences.edit();
        preferencesEditor.putBoolean(getString(R.string.preference_name_rating_dont_show_rating_prompt), true);
        preferencesEditor.apply();
        setVisibility(R.id.rating_frame_layout, View.INVISIBLE);
        setVisibility(R.id.text_warning, View.VISIBLE);
        Intent sendEmailIntent = new Intent(Intent.ACTION_SENDTO);
        String uriText = "mailto:" + Uri.encode(getString(R.string.feedback_email_address)) + "?subject=" + Uri.encode("Circuit & Outlet Tester Feedback");
        Uri uri = Uri.parse(uriText);
        sendEmailIntent.setData(uri);
        startActivity(Intent.createChooser(sendEmailIntent, "Send e-mail..."));
    }

    // when a user clicks the "no rating" rating button, hide the rating prompt
    public void onRatingNoRatingButtonClicked(View view){
        SharedPreferences preferences = this.getSharedPreferences(getString(R.string.preference_shared_key_name_rating), 0);
        SharedPreferences.Editor preferencesEditor = preferences.edit();
        preferencesEditor.putBoolean(getString(R.string.preference_name_rating_dont_show_rating_prompt), true);
        preferencesEditor.apply();
        setVisibility(R.id.rating_frame_layout, View.INVISIBLE);
        setVisibility(R.id.text_warning, View.VISIBLE);
    }

    // when a user clicks the "yes rating" rating button, hide the prompt and open google play
    public void onRatingYesRatingButtonClicked(View view){
        SharedPreferences preferences = this.getSharedPreferences(getString(R.string.preference_shared_key_name_rating), 0);
        SharedPreferences.Editor preferencesEditor = preferences.edit();
        preferencesEditor.putBoolean(getString(R.string.preference_name_rating_dont_show_rating_prompt), true);
        preferencesEditor.apply();
        setVisibility(R.id.rating_frame_layout, View.INVISIBLE);
        setVisibility(R.id.text_warning, View.VISIBLE);

        // try launching the google play app, otherwise fall back to the web browser
        try {
            Intent playStoreRatingIntent = new Intent(Intent.ACTION_VIEW, Uri.parse("market://details?id=" + getString(R.string.app_package_name)));
            startActivity(playStoreRatingIntent);
        } catch (android.content.ActivityNotFoundException exception){
            Intent googlePlayBrowserIntent = new Intent(Intent.ACTION_VIEW, Uri.parse("https://play.google.com/store/apps/details?id=" + getString(R.string.app_package_name)));
            startActivity(googlePlayBrowserIntent);
        }
    }
}

File: mobile/src/main/java/xyz/mechenbier/circuittester/MainActivity.kt
package xyz.mechenbier.circuittester

import android.app.*
import android.content.*
import android.net.Uri
import android.os.Bundle
import android.os.IBinder
import androidx.preference.PreferenceManager
import androidx.core.content.ContextCompat
import androidx.appcompat.app.AppCompatActivity
import android.util.Log
import android.util.TypedValue
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.*
import com.google.android.gms.ads.AdView
import com.google.android.gms.ads.AdRequest
import com.google.android.gms.ads.MobileAds
import com.google.firebase.analytics.FirebaseAnalytics

class MainActivity : AppCompatActivity() {

    private lateinit var mService: PowerStateService
    private var mBound = false
    private var mUiUpdateReceiver: UiUpdateReceiver? = null
    private lateinit var mFirebaseAnalytics: FirebaseAnalytics
    private val mRatingDaysUntilPrompt = 3
    private val mRatingLaunchesUntilPrompt = 3

    companion object {
        lateinit var mInstance: MainActivity
    }

    //Defines callbacks for service binding, passed to bindService()
    private val mConnection = object : ServiceConnection {

        override fun onServiceConnected(className: ComponentName, service: IBinder) {
            // We've bound to LocalService, cast the IBinder and get LocalService instance
            val binder = service as PowerStateService.LocalBinder
            mService = binder.getService()
            mBound = true

            val preferences = PreferenceManager.getDefaultSharedPreferences(applicationContext)
            val prefValueSoundWhenPowered: Boolean = preferences.getBoolean(KEY_PREF_SOUND_WHEN_POWERED, false)
            val prefValueSoundMute: Boolean = preferences.getBoolean(KEY_PREF_SOUND_MUTE, false)

            mService.setSoundOnPowered(prefValueSoundWhenPowered)
            mService.setMute(prefValueSoundMute)
        }

        override fun onServiceDisconnected(arg0: ComponentName) {
            mBound = false
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        PreferenceManager.setDefaultValues(this, R.xml.preferences, false)

        mInstance = this
        mFirebaseAnalytics = FirebaseAnalytics.getInstance(this)

        MobileAds.initialize(this)

        val adView = findViewById<View>(R.id.adView) as AdView
        val adRequest = AdRequest.Builder()
                .addTestDevice(AdRequest.DEVICE_ID_EMULATOR)
                .build()
        adView.loadAd(adRequest)

        val preferences = PreferenceManager.getDefaultSharedPreferences(this)

        initializeUiFromPreferences(preferences)
        initializeRatingPrompt()
        startService(preferences)
    }


    override fun onStart() {
        super.onStart()
        bindService()
    }

    override fun onResume() {
        super.onResume()

        if (mUiUpdateReceiver == null) {
            mUiUpdateReceiver = UiUpdateReceiver()
        }

        val intentFilter = android.content.IntentFilter(INTENT_POWER_CONNECTION_RECEIVER_CHARGING)
        registerReceiver(mUiUpdateReceiver, intentFilter)

        val preferences = PreferenceManager.getDefaultSharedPreferences(this)
        initializeUiFromPreferences(preferences)
    }

    override fun onPause() {
        if (mUiUpdateReceiver != null) {
            unregisterReceiver(mUiUpdateReceiver)
        }
        super.onPause()
    }

    override fun onStop() {
        super.onStop()
        unbindService()
    }

    override fun onDestroy() {
        stopService()
        super.onDestroy()
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        val inflater = getMenuInflater()
        inflater.inflate(R.menu.menu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when(item.itemId){
            R.id.menu_action_help -> {

                val builder = AlertDialog.Builder(this)
                builder.setMessage(R.string.help_dialog_message)
                        .setNeutralButton(R.string.launch_feedback_text) { dialogInterface, i ->
                            dialogInterface.cancel()
                            launchSupportEmailIntent()
                        }
                        .setNegativeButton(R.string.launch_privacy_policy_text) { dialogInterface, i ->
                            dialogInterface.cancel()
                            launchPrivacyPolicyIntent()
                        }
                        .setPositiveButton(R.string.help_dialog_close_button_text) { dialogInterface, i ->
                            dialogInterface.cancel()
                        }
                val alertDialog = builder.create()
                alertDialog.show()

                true
            }
            R.id.menu_action_settings -> {

                val intent = Intent(this, SettingsActivity::class.java)
                startActivity(intent)

                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    private fun bindService() {
        // Bind to LocalService
        val intent = Intent(this, PowerStateService::class.java)
        bindService(intent, mConnection, Context.BIND_AUTO_CREATE)
    }

    private fun unbindService() {
        try {
            unbindService(mConnection)
        } catch (e: Exception) {
            Log.e("SERVICE", "Issue unbinding the service")
        }

        mBound = false
    }

    private fun startService(preferences: SharedPreferences) {
        val prefValueSoundWhenPowered: Boolean = preferences.getBoolean(KEY_PREF_SOUND_WHEN_POWERED, false)
        val prefValueSoundMute: Boolean = preferences.getBoolean(KEY_PREF_SOUND_MUTE, false)
        val powerStateServiceIntent = Intent(this, PowerStateService::class.java)

        powerStateServiceIntent.putExtra(POWER_STATE_SERVICE_INTENT_EXTRA_IS_MUTED, prefValueSoundMute)
        powerStateServiceIntent.putExtra(POWER_STATE_SERVICE_INTENT_EXTRA_SOUND_WHEN_POWERED, prefValueSoundWhenPowered)

        startService(powerStateServiceIntent)
    }

    private fun stopService() {
        unbindService()
        val intent = Intent(this, PowerStateService::class.java)
        stopService(intent)
    }


    fun onRadioButtonClicked(view: View) {
        // Check which radio button was clicked
        when (view.getId()) {
            R.id.rb_sound_when_powered -> setSoundOnPowered(true)
            R.id.rb_sound_when_not_powered -> setSoundOnPowered(false)
        }
    }

    fun onMuteToggleButtonClicked(view: View) {
        val isChecked = (view as ToggleButton).isChecked
        setMute(isChecked)
    }


    private fun setSoundOnPowered(checked: Boolean) {
        val sharedPref: SharedPreferences = PreferenceManager.getDefaultSharedPreferences(this)
        val editor: SharedPreferences.Editor = sharedPref.edit()
        editor.putBoolean(KEY_PREF_SOUND_WHEN_POWERED, checked)
        editor.apply()

        if (mBound) {
            mService!!.setSoundOnPowered(checked)
        }
    }

    private fun setMute(muted: Boolean) {
        val sharedPref: SharedPreferences = PreferenceManager.getDefaultSharedPreferences(this)
        val editor: SharedPreferences.Editor = sharedPref.edit()
        editor.putBoolean(KEY_PREF_SOUND_MUTE, muted)
        editor.apply()

        if (mBound) {
            mService!!.setMute(muted)
        }
    }


    fun setUiFromChargingState(isCharging: Boolean) {
        val image = findViewById<ImageView>(R.id.image_powerstate)
        val powerStateTextView = findViewById<TextView>(R.id.text_powerstate)

        if (isCharging) {
            image.setColorFilter(ContextCompat.getColor(this, R.color.image_charging))
            powerStateTextView.text = getString(R.string.power_state_powered)
        } else {
            image.setColorFilter(ContextCompat.getColor(this, R.color.image_not_charging))
            powerStateTextView.text = getString(R.string.power_state_not_powered)
        }
    }

    private fun initializeUiFromPreferences(preferences: SharedPreferences) {
        val prefValueShowPowerStateText = preferences.getBoolean(KEY_PREF_SHOW_STATE_TEXT, false)
        val prefValueSoundWhenPowered = preferences.getBoolean(KEY_PREF_SOUND_WHEN_POWERED, false)
        val prefValueSoundMute = preferences.getBoolean(KEY_PREF_SOUND_MUTE, false)

        val tbMute = findViewById<ToggleButton>(R.id.toggle_mute)
        val rbSoundWhenPowered = findViewById<RadioButton>(R.id.rb_sound_when_powered)
        val rbSoundWhenNotPowered = findViewById<RadioButton>(R.id.rb_sound_when_not_powered)
        val powerStateTextView = findViewById<TextView>(R.id.text_powerstate)
        val powerStateTextViewParams = powerStateTextView.layoutParams

        tbMute.isChecked = prefValueSoundMute
        rbSoundWhenPowered.isChecked = prefValueSoundWhenPowered
        rbSoundWhenNotPowered.isChecked = !prefValueSoundWhenPowered

        if (prefValueShowPowerStateText) {
            powerStateTextViewParams.height = RelativeLayout.LayoutParams.WRAP_CONTENT
            setVisibility(R.id.text_powerstate, View.VISIBLE)
        } else {
            val viewHeight = TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 0.toFloat(), getResources().getDisplayMetrics()).toInt()
            powerStateTextViewParams.height = viewHeight
            setVisibility(R.id.text_powerstate, View.INVISIBLE)
        }

        powerStateTextView.layoutParams = powerStateTextViewParams
        powerStateTextView.requestLayout()
    }

    private fun initializeRatingPrompt() {
        val preferences = this.getSharedPreferences(KEY_PREF_SHARED_RATING_KEY_NAME, 0)

        if (preferences.getBoolean(KEY_PREF_RATING_DONT_SHOW_RATING_PROMPT, false)) {
            return
        }

        val preferencesEditor = preferences.edit()

        // increment launch counter
        val launchCount = preferences.getLong(KEY_PREF_RATING_LAUNCH_COUNT, 0) + 1
        preferencesEditor.putLong(KEY_PREF_RATING_LAUNCH_COUNT, launchCount)

        // get date of first launch
        var dateFirstLaunch = preferences.getLong(KEY_PREF_RATING_DATE_FIRST_LAUNCH, 0)
        if (dateFirstLaunch == 0L) {
            dateFirstLaunch = System.currentTimeMillis()
            preferencesEditor.putLong(KEY_PREF_RATING_DATE_FIRST_LAUNCH, dateFirstLaunch)
        }

        preferencesEditor.apply()

        // wait at least n days before opening
        if (launchCount >= mRatingLaunchesUntilPrompt) {
            if (System.currentTimeMillis() >= dateFirstLaunch + mRatingDaysUntilPrompt * 24 * 60 * 60 * 1000) {
                showHideRatingView(true)
                showHideWarningText(false)
            }
        }
    }

    private fun setVisibility(id: Int, visibility: Int) {
        val view = findViewById<View>(id)
        if (view != null) {
            view.visibility = visibility
        }
    }

    // when a user clicks the "not enjoying" rating button, hide the enjoyment prompts and show the feedback prompts
    fun onRatingNotEnjoyingButtonClicked(view: View) {
        setVisibility(R.id.enjoying_circuit_tester, View.INVISIBLE)
        setVisibility(R.id.button_enjoying_no, View.INVISIBLE)
        setVisibility(R.id.button_enjoying_yes, View.INVISIBLE)
        setVisibility(R.id.feedback_circuit_tester, View.VISIBLE)
        setVisibility(R.id.button_feedback_no, View.VISIBLE)
        setVisibility(R.id.button_feedback_yes, View.VISIBLE)
    }

    // when a user clicks the "yes enjoying" rating button, hide the enjoyment prompts and show the rating prompts
    fun onRatingYesEnjoyingButtonClicked(view: View) {
        setVisibility(R.id.enjoying_circuit_tester, View.INVISIBLE)
        setVisibility(R.id.button_enjoying_no, View.INVISIBLE)
        setVisibility(R.id.button_enjoying_yes, View.INVISIBLE)
        setVisibility(R.id.rate_circuit_tester, View.VISIBLE)
        setVisibility(R.id.button_rating_no, View.VISIBLE)
        setVisibility(R.id.button_rating_yes, View.VISIBLE)
    }

    // when a user clicks the "no feedback" rating button, hide the rating prompt
    fun onRatingNoFeedbackButtonClicked(view: View) {
        setDontShowRatingPreference(true)
        showHideRatingView(false)
        showHideWarningText(true)
    }

    // when a user clicks the "yes feedback" rating button, hide the prompt and create an email intent
    fun onRatingYesFeedbackButtonClicked(view: View) {
        setDontShowRatingPreference(true)
        showHideRatingView(false)
        showHideWarningText(true)

        launchSupportEmailIntent()
    }

    // when a user clicks the "no rating" rating button, hide the rating prompt
    fun onRatingNoRatingButtonClicked(view: View) {
        setDontShowRatingPreference(true)
        showHideRatingView(false)
        showHideWarningText(true)
    }

    // when a user clicks the "yes rating" rating button, hide the prompt and open google play
    fun onRatingYesRatingButtonClicked(view: View) {
        setDontShowRatingPreference(true)
        showHideRatingView(false)
        showHideWarningText(true)

        launchGooglePlayIntent()
    }

    private fun setDontShowRatingPreference(dontShow: Boolean){
        val preferences = this.getSharedPreferences(KEY_PREF_SHARED_RATING_KEY_NAME, 0)
        val preferencesEditor = preferences.edit()
        preferencesEditor.putBoolean(KEY_PREF_RATING_DONT_SHOW_RATING_PROMPT, dontShow)
        preferencesEditor.apply()
    }

    private fun showHideRatingView(show: Boolean){
        val ratingView = findViewById<View>(R.id.rating_frame_layout)
        val ratingParams = ratingView.layoutParams

        if (show) {
            val ratingHeight = TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 115.toFloat(), getResources().getDisplayMetrics()).toInt()
            ratingParams.height = ratingHeight
            setVisibility(R.id.rating_frame_layout, View.VISIBLE)
        } else {
            val ratingHeight = TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 0.toFloat(), getResources().getDisplayMetrics()).toInt()
            ratingParams.height = ratingHeight
            setVisibility(R.id.rating_frame_layout, View.INVISIBLE)
        }

        ratingView.layoutParams = ratingParams
        ratingView.requestLayout()
    }

    private fun showHideWarningText(show: Boolean){
        val warningView = findViewById<View>(R.id.text_warning)
        val warningParams = warningView.layoutParams

        if (show){
            warningParams.height = RelativeLayout.LayoutParams.WRAP_CONTENT
            setVisibility(R.id.text_warning, View.VISIBLE)
        } else {
            val warningHeight = TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 0.toFloat(), getResources().getDisplayMetrics()).toInt()
            warningParams.height = warningHeight
            setVisibility(R.id.text_warning, View.INVISIBLE)
        }

        warningView.layoutParams = warningParams
        warningView.requestLayout()
    }

    private fun launchGooglePlayIntent(){
        // try launching the google play app, otherwise fall back to the web browser
        try {
            val playStoreRatingIntent = Intent(Intent.ACTION_VIEW, Uri.parse("market://details?id=xyz.mechenbier.circuittester"))
            startActivity(playStoreRatingIntent)
        } catch (exception: android.content.ActivityNotFoundException) {
            val googlePlayBrowserIntent = Intent(Intent.ACTION_VIEW, Uri.parse("https://play.google.com/store/apps/details?id=xyz.mechenbier.circuittester"))
            startActivity(Intent.createChooser(googlePlayBrowserIntent, getString(R.string.intent_title_launch_browser)))
        }
    }

    private fun launchSupportEmailIntent(){
        val sendEmailIntent = Intent(Intent.ACTION_SENDTO)
        val uriText = "mailto:" + Uri.encode(getString(R.string.feedback_email_address)) + "?subject=" + Uri.encode(getString(R.string.app_name) + " " + getString(R.string.launch_feedback_text))
        val uri = Uri.parse(uriText)
        sendEmailIntent.data = uri
        startActivity(Intent.createChooser(sendEmailIntent, getString(R.string.intent_title_send_email)))
    }

    private fun launchPrivacyPolicyIntent(){
        val openPrivacyPolicyIntent = Intent(Intent.ACTION_VIEW, Uri.parse(getString(R.string.privacy_policy_url)))
        startActivity(Intent.createChooser(openPrivacyPolicyIntent, getString(R.string.intent_title_launch_browser)))
    }
}